### PR TITLE
Vector

### PR DESCRIPTION
**Pull request by EgorKalinniKov**

### CODE REVIEW

File: Vector.cpp [hunk 1]
```diff
@@ -0,0 +1,238 @@
...
+    explicit Custom_Vector(size_t size)
+        : _size(size),
+          _capacity(size),
+          buffer(new int[_capacity])
+    {
+        std::fill_n(buffer, _size, int());
+    }
+
+    Custom_Vector(const std::initializer_list<int> list)
+        : Custom_Vector(list.size())
```
Comment (bduminiuc) [inline Vector.cpp]: Think about what happens here. The vector must be created with the list of values. The code redirects the creation to another constructor. It initializes the vector, fills it with default values, and then refills it with the values from the list.

It looks like a filling logic duplication. C++ is about efficiency, we must avoid unnecessary operations where possible.

By the way, the constructor with `initializer_list` does not contain `const` modifier on cppreference.

File: Vector.cpp [hunk 2]
```diff
@@ -0,0 +1,238 @@
...
+    {
+        std::copy(list.begin(), list.end(), buffer);
+    }
+
+    Custom_Vector(const Custom_Vector &other)
+        : Custom_Vector(other._size)
```
Comment (bduminiuc) [inline Vector.cpp]: The same thing with this constructor. Refilling of the vector happens again.

File: Vector.cpp [hunk 3]
```diff
@@ -0,0 +1,238 @@
...
+    {
+        std::copy(other.buffer, other.buffer + other._size, buffer);
+    }
+
+    Custom_Vector(Custom_Vector &&other)
+        : Custom_Vector()
```
Comment (bduminiuc) [inline Vector.cpp]: Similar thing here. The current implementation initializes `size` and `capacity` with zeros and then swaps. You can initialize them directly with the proper values. You need only buffer swapping because it is heap-allocated memory.

File: Vector.cpp [hunk 4]
```diff
@@ -0,0 +1,238 @@
...
+        reserve(count);
+
+        _size = count;
+
+        std::fill_n(buffer, _size, value);
+    }
+
+    void assign(std::initializer_list<int> list)
+    {
+        clear();
```
Comment (bduminiuc) [inline Vector.cpp]: The `reserve` implementation does not refer to `size`. So you don't need to reset it.

File: Vector.cpp [hunk 5]
```diff
@@ -0,0 +1,238 @@
...
+        {
+            return;
+        }
+
+        int *point = new int[size_n];
+
+        std::copy(buffer, buffer + _size, point);
+
+        delete[] buffer;
+        std::swap(buffer, point);
```
Comment (bduminiuc) [inline Vector.cpp]: You can only assign `buffer = point` instead of swap because swapping is less efficient.

File: Vector.cpp [hunk 6]
```diff
@@ -0,0 +1,238 @@
...
+        _capacity = size_n;
+    }
+
+    int at(size_t pos)
+    {
+        if (_size - 1 < pos || _size == 0)
```
Comment (bduminiuc) [inline Vector.cpp]: When `_size - 1` is executed and the vector is empty the behavior is undefined. You should check emptiness before. Actually, cppreference notes the correct condition when the exception must be thrown. 

You also have this condition in other places. Check all of them, please.

File: Vector.cpp [hunk 7]
```diff
@@ -0,0 +1,238 @@
...
+            return *this;
+        }
+
+        int *point = new int[other._capacity];
+        std::copy(other.buffer, other.buffer + other._size, point);
+
+        delete[] buffer;
+        _size = other._size;
+        _capacity = other._capacity;
+        std::swap(buffer, point);
```
Comment (bduminiuc) [inline Vector.cpp]: assign instead of swap

File: Vector.cpp [hunk 8]
```diff
@@ -0,0 +1,238 @@
...
+        else
+        {
+            return buffer[pos];
+        }
+    }
+
+    int &operator[](size_t pos) { return buffer[pos]; }
+    const int &operator[](size_t pos) const { return buffer[pos]; }
+
+    Custom_Vector &operator=(const Custom_Vector &other)
```
Comment (bduminiuc) [inline Vector.cpp:144]: `operator=(Custom_Vector &&other)` is missed

File: Vector.cpp [hunk 9]
```diff
@@ -22,24 +27,29 @@ class Custom_Vector
...
     Custom_Vector(Custom_Vector &&other)
-        : Custom_Vector()
+        : _size(other._size),
+          _capacity(other._capacity),
+          buffer(nullptr)
     {
         std::swap(buffer, other.buffer);
-        std::swap(_size, other._size);
-        std::swap(_capacity, other._capacity);
+        other._size = other._capacity = 0;
```
Comment (bduminiuc) [inline Vector.cpp:52]: No need to reset `other` fields because it is the "moved from" object here and in `operator=(Custom_Vector &&other)`.

File: Vector.cpp [hunk 10]
```diff
@@ -173,66 +201,224 @@ class Custom_Vector
...
         _size--;
     }
+
+    iterator begin() { return buffer; }
+    const_iterator begin() const { return buffer; }
+
+    iterator end() { return buffer + _size; }
+    const_iterator end() const { return buffer + _size; }
+
+    iterator cbegin() { return buffer + _size - 1; }
```
Comment (bduminiuc) [inline Vector.cpp]: why do `begin`/`cbegin` implementations differ? The same is about `end`/`cend`. They must be equal

File: Vector.cpp [hunk 11]
```diff
@@ -173,66 +201,224 @@ class Custom_Vector
...
-    assert(c.size() == 0);
-    assert(c.capacity() == 6);
-    return 0;
+    Custom_Vector test;
+
+    EXPECT_EQ(test.size(), 0);
+    EXPECT_EQ(test.capacity(), 0);
+}
+
+TEST(Vector_Test, constructorBySize)
```
Comment (bduminiuc) [inline Vector.cpp:254]: The following cases must be covered:
- input is empty
- input is not empty

Check the values here. You can write both in the same `TEST`.

Also, verify other tests cover empty case.

File: Vector.cpp [hunk 12]
```diff
@@ -173,66 +201,224 @@ class Custom_Vector
...
+    std::initializer_list<int> tmp = {1, 2, 3, 4};
+    Custom_Vector test1(tmp);
+    std::vector<int> test2(tmp);
+
+    EXPECT_EQ(test1.size(), test2.size());
+    EXPECT_EQ(test1.capacity(), test2.capacity());
+    EXPECT_TRUE(std::equal(test1.begin(), test1.end(), test2.begin()));
+}
+
+TEST(Vector_Test, copyConstructor)
```
Comment (bduminiuc) [inline Vector.cpp:305]: In the copy constructor, you must also verify the destination object is the copy of the source one. You can add the `operator==` implementation in `Custom_Vector`. In the implementation, you must check the equality of sizes, capacities, and buffer elements.

File: Vector.cpp [hunk 13]
```diff
@@ -173,66 +201,224 @@ class Custom_Vector
...
+    Custom_Vector test1;
+
+    EXPECT_THROW(test1.at(0), std::out_of_range);
+    EXPECT_THROW(test1.at(-1), std::out_of_range);
+    EXPECT_THROW(test1.at(1), std::out_of_range);
+
+    const Custom_Vector test2 = {1, 2, 3, 4, 5};
+    EXPECT_THROW(test2.at(-1), std::out_of_range);
+    EXPECT_THROW(test2.at(5), std::out_of_range);
+    EXPECT_NO_THROW(test2.at(1));
```
Comment (bduminiuc) [inline Vector.cpp]: The value must be also equal to 2.

File: Vector.cpp [hunk 14]
```diff
@@ -173,66 +201,224 @@ class Custom_Vector
...
+    EXPECT_EQ(test1.capacity(), test2.capacity());
+}
+
+TEST(Vector_Test, constructorByInitList)
+{
+    std::initializer_list<int> tmp = {1, 2, 3, 4};
+    Custom_Vector test1(tmp);
+    std::vector<int> test2(tmp);
+
+    EXPECT_EQ(test1.size(), test2.size());
```
Comment (bduminiuc) [inline Vector.cpp]: To simplify the tests logic you can add free helper functions like
```C++
bool equivalent(const Custom_Vector& lhs, const std::vector<int> &rhs) {
    EXPECT_EQ(lhs.size(), rhs.size());
    EXPECT_TRUE(lhs.capacity() <= rhs.capacity());
    EXPECT_TRUE(std::equal(lhs.begin(), lhs.end(), rhs.begin()));
}

bool operator==(const Custom_Vector& lhs, const std::vector<int> &rhs) {
    EXPECT_EQ(lhs.size(), rhs.size());
    EXPECT_EQ(lhs.capacity(), rhs.capacity());
    EXPECT_TRUE(std::equal(lhs.begin(), lhs.end(), rhs.begin()));
}
```
And use them instead. The code will become shorter.
